audio/firmware.c

```diff
@@ -1,201 +1,129 @@
-#include <fx2macros.h>
-#include <serial.h>
-#include <delay.h>
 #include <autovector.h>
-#include <setupdat.h>
+#include <delay.h>
 #include <eputils.h>
+#include <fx2macros.h>
+#include <fx2regs.h>
+#include <fx2types.h>
+#include <setupdat.h>
 
 #include "fx2lights.h"
+#include "audiodata.h"
 
 #define SYNCDELAY SYNCDELAY4
-#define REARMVAL 0x80
-#define REARM() EP2BCL=REARMVAL
 
-volatile WORD bytes;
-volatile __bit gotbuf;
-volatile BYTE icount;
 volatile __bit got_sud;
-DWORD lcount;
-__bit on;
+extern BYTE alt_setting;
 
 void main() {
-    REVCTL=0; // not using advanced endpoint controls
+    d1on();
+    d2on();
+    /* Not using advanced endpoint controls */
+    REVCTL=0;
 
-    d1off();
-    on=0;
-    lcount=0;
-    got_sud=FALSE;
-    icount=0;
-    gotbuf=FALSE;
-    bytes=0;
+    got_sud = FALSE;
 
-    // renumerate
-    RENUMERATE_UNCOND(); 
+    /* renumerate */
+    RENUMERATE_UNCOND();
 
     SETCPUFREQ(CLK_48M);
     SETIF48MHZ();
-    sio0_init(57600);
+    usart_init();
 
-    USE_USB_INTS(); 
+    USE_USB_INTS();
     ENABLE_SUDAV();
-    ENABLE_SOF();
-    ENABLE_HISPEED();
     ENABLE_USBRESET();
+    ENABLE_HISPEED();
+    d1off();
 
-    // only valid endpoints is 2
-    EP2CFG = 0xA2; // 10100010
-    SYNCDELAY;
-    EP1OUTCFG &= ~bmVALID;
-    SYNCDELAY;
-    EP4CFG &= ~bmVALID;
-    SYNCDELAY;
-
-    // arm ep2
-    EP2BCL = 0x80; // write once
-    SYNCDELAY;
-    EP2BCL = 0x80; // do it again
-
-    // make it so we enumerate
-    EA=1; // global interrupt enable 
-    printf ( "Done initializing stuff\n" );
-
+    /**
+     * No valid endpoints by default, thus clear the valid bit and set the
+     * reset to default
+     * Invalid
+     * Type: Bulk
+     * Direction: out
+     * Buffer: Double
+     */
+    EP2CFG = 0x7F;
+    SYNCDELAY; EP4CFG = 0x7F;
+    SYNCDELAY; EP6CFG = 0x7F;
+    SYNCDELAY; EP8CFG = 0x7F;
+
+    /* Enable global interrupts */
+    EA = 1;
+    /* Disable serial interrupts */
+    ES0 = 0;
     d2off();
 
+    printf("Initialisation complete\n");
+
     while(TRUE) {
         if (got_sud) {
-            printf ( "Handle setupdata\n" );
-            handle_setupdata(); 
-            got_sud=FALSE;
+            printf("Handle setup data\n");
+            handle_setupdata();
+            got_sud = FALSE;
         }
-
-        if (!(EP2468STAT & bmEP2EMPTY)) {
-            printf ( "ep2 out received data\n" );
-            if (!(EP2468STAT & bmEP6FULL)) { // wait for at least one empty in buffer
-                WORD i;
-                printf ( "Sending data to ep6 in\n");
-
-                bytes = MAKEWORD(EP2BCH,EP2BCL);
-
-                for (i=0;i<bytes;++i) EP6FIFOBUF[i] = EP2FIFOBUF[i];
-
-                // can copy whole string w/ autoptr instead.
-                // or copy directly from one buf to another
-
-                // ARM ep6 out
-                EP6BCH=MSB(bytes);
+        /* ISO endpoint config type is 01 in the enpoint configuration buffer */
```

Comment:
Why was the iso endpoint structure empty in the descriptors?

Reply:
It isn't...?